servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ReplayStrategies.java

```diff
@@ -0,0 +1,382 @@
+                () -> new LazyTimeLimitedReplayAccumulator<>(historyHint, ttl, executor) :
+                () -> new EagerTimeLimitedReplayAccumulator<>(historyHint, ttl, executor));
+    }
+
+    private static final class MostRecentReplayAccumulator<T> implements ReplayAccumulator<T> {
+        private final int maxItems;
+        private final Deque<Object> queue = new ArrayDeque<>();
+
+        MostRecentReplayAccumulator(final int maxItems) {
+            if (maxItems <= 0) {
+                throw new IllegalArgumentException("maxItems: " + maxItems + "(expected >0)");
+            }
+            this.maxItems = maxItems;
+        }
+
+        @Override
+        public void accumulate(@Nullable final T t) {
+            if (queue.size() >= maxItems) {
+                queue.poll();
+            }
+            queue.add(wrapNull(t));
+        }
+
+        @Override
+        public void deliverAccumulation(final Consumer<T> consumer) {
+            for (Object item : queue) {
+                consumer.accept(unwrapNullUnchecked(item));
+            }
+        }
+    }
+
+    private static final class LazyTimeLimitedReplayAccumulator<T> implements ReplayAccumulator<T> {
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<LazyTimeLimitedReplayAccumulator> queueSizeUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(LazyTimeLimitedReplayAccumulator.class, "queueSize");
+        private final Executor executor;
+        private final Queue<TimeStampSignal<T>> items;
+        private final long ttlNanos;
+        private final int maxItems;
+        private volatile int queueSize;
+
+        LazyTimeLimitedReplayAccumulator(final int maxItems, final Duration ttl, final Executor executor) {
+            if (ttl.isNegative()) {
+                throw new IllegalArgumentException("ttl: " + ttl + "(expected non-negative)");
+            }
+            if (maxItems <= 0) {
+                throw new IllegalArgumentException("maxItems: " + maxItems + "(expected >0)");
+            }
+            this.executor = requireNonNull(executor);
+            this.ttlNanos = ttl.toNanos();
+            this.maxItems = maxItems;
+            items = new ConcurrentLinkedQueue<>(); // SpMc
+        }
+
+        @Override
+        public void accumulate(@Nullable final T t) {
+            final TimeStampSignal<T> signal = new TimeStampSignal<>(executor.currentTime(NANOSECONDS), t);
+            for (;;) {
+                final int qSize = queueSize;
+                if (qSize < maxItems) {
+                    if (queueSizeUpdater.compareAndSet(this, qSize, qSize + 1)) {
+                        items.add(signal);
+                        break;
+                    }
+                } else if (queueSizeUpdater.compareAndSet(this, qSize, qSize)) {
+                    items.poll();
+                    items.add(signal);
+                    break;
+                }
+            }
+        }
```

Comment:
This is also an opportunity to evict stale elements if we'd like. We can have a simple `private evictStale()` method and call it at the start of these public methods. Then the rest looks almost exactly like `MostRecentReplayAccumulator`.

Reply:
lets defer this bcz `LazyTimeLimitedReplayAccumulator` can do the iteration and removal at the same time which is a bit harder to generalize.